File: windows/tests/RealtimeTranslator.Core.Tests/RealtimeSubtitleAssemblerAudioLossTests.cs
using System;
using RealtimeTranslator.Core.OpenAI;
using RealtimeTranslator.Core.Realtime;
using Xunit;

namespace RealtimeTranslator.Core.Tests;

public sealed class RealtimeSubtitleAssemblerAudioLossTests
{
    private static readonly DateTimeOffset Origin =
        new(2026, 1, 1, 0, 0, 0, TimeSpan.Zero);

    // Given: 音声欠落を検知して汚染窓を開始する
    // When: 原文と訳文を受け取り、idle finalize を評価する
    // Then: 確定せず、無効化更新で汚染セグメントのバッファを破棄する
    [Fact]
    public void TaintedSegmentIsAbandonedInsteadOfFinalized()
    {
        var assembler = NewAssembler();
        assembler.MarkAudioLoss(Origin);
        assembler.Ingest(Source("こんにちは", "s1"), Origin.AddMilliseconds(1));
        assembler.Ingest(Translation("Hello", "t1"), Origin.AddMilliseconds(2));

        var update = assembler.Tick(Origin.AddSeconds(9));

        Assert.NotNull(update);
        Assert.True(update.Value.IsInvalidation);
        Assert.False(update.Value.ShouldFinalize);
        Assert.Equal(string.Empty, assembler.CurrentSourceText);
        Assert.False(assembler.IsCurrentSegmentTainted);
    }

    // Given: 音声欠落の汚染窓が期限切れになっている
    // When: 期限後に新しい原文と訳文を受け取り idle finalize する
    // Then: 新しいセグメントは通常どおり確定する
    [Fact]
    public void SegmentAfterTaintWindowCanFinalize()
    {
        var assembler = NewAssembler();
        assembler.MarkAudioLoss(Origin);
        assembler.Ingest(
            Source("ありがとう", "s1"),
            Origin.Add(RealtimeSubtitleAssembler.AudioLossTaintWindow).AddMilliseconds(1));
        assembler.Ingest(Translation("Thank you", "t1"), Origin.AddSeconds(8.2));

        var update = assembler.Tick(Origin.AddSeconds(17));

        Assert.NotNull(update);
        Assert.True(update.Value.ShouldFinalize);
        Assert.Equal("ありがとう", update.Value.SourceText);
        Assert.Equal("Thank you", update.Value.TranslatedText);
    }

    // Given: 音声欠落より前に字幕ペアが確定している
    // When: その後に音声欠落を通知する
    // Then: 既に確定した更新は変更されない
    [Fact]
    public void FinalizedSegmentIsPreservedAcrossAudioLoss()
    {
        var assembler = NewAssembler();
        assembler.Ingest(Source("こんにちは", "s1"), Origin);
        assembler.Ingest(Translation("Hello", "t1"), Origin.AddMilliseconds(1));
        var finalized = assembler.Tick(Origin.AddSeconds(9));

        assembler.MarkAudioLoss(Origin.AddSeconds(10));

        Assert.NotNull(finalized);
        Assert.True(finalized.Value.ShouldFinalize);
        Assert.Equal("こんにちは", finalized.Value.SourceText);
        Assert.Equal("Hello", finalized.Value.TranslatedText);
    }

    // Given: 音声欠落後の汚染セグメントに完全なペアがある
    // When: 言語切替で分割する
    // Then: 汚染された prefix は確定しない
    [Fact]
    public void TaintedLanguageSwitchDoesNotFinalizePrefix()
    {
        var assembler = NewAssembler();
        assembler.MarkAudioLoss(Origin);
        assembler.Ingest(Source("こんにちは", "s1"), Origin.AddMilliseconds(1));
        assembler.Ingest(Translation("Hello", "t1"), Origin.AddMilliseconds(2));

        var split = assembler.SplitForLanguageSwitch(3, Origin.AddMilliseconds(3));

        Assert.Null(split.Finalized);
    }

    // Given: 汚染された原文の途中に言語切替境界がある
    // When: suffix を残したあと新しい訳文を受け取り idle finalize する
    // Then: 汚染 suffix は確定せず無効化される
    [Fact]
    public void TaintedLanguageSwitchKeepsSuffixFromFinalizing()
    {
        var assembler = NewAssembler();
        assembler.MarkAudioLoss(Origin);
        assembler.Ingest(Source("こんにちはHello", "s1"), Origin.AddMilliseconds(1));
        assembler.Ingest(Translation("Hello", "t1"), Origin.AddMilliseconds(2));

        var split = assembler.SplitForLanguageSwitch(5, Origin.AddMilliseconds(3));

        Assert.Null(split.Finalized);
        Assert.Equal("Hello", assembler.CurrentSourceText);
        Assert.True(assembler.IsCurrentSegmentTainted);

        assembler.ExpectLane(RealtimeTranslationOutputLanguage.Japanese);
        assembler.Ingest(
            new RealtimeTranslationStreamEvent(
                RealtimeTranslationLane.Translation(RealtimeTranslationOutputLanguage.Japanese),
                new RealtimeTranslationServerEvent.OutputTranscriptDelta("こんにちは", "t2", 300),
                1),
            Origin.AddSeconds(1));
        var update = assembler.Tick(Origin.AddSeconds(10));

        Assert.NotNull(update);
        Assert.True(update.Value.IsInvalidation);
        Assert.False(update.Value.ShouldFinalize);
        Assert.Equal(string.Empty, assembler.CurrentSourceText);
        Assert.False(assembler.IsCurrentSegmentTainted);
    }

    private static RealtimeSubtitleAssembler NewAssembler()
    {
        var assembler = new RealtimeSubtitleAssembler();
        assembler.Reset(1);
        assembler.ExpectLane(RealtimeTranslationOutputLanguage.English);
        return assembler;
    }

    private static RealtimeTranslationStreamEvent Source(string text, string eventId) =>
        new(
            RealtimeTranslationLane.Source,
            new RealtimeTranslationServerEvent.InputTranscriptDelta(text, eventId, 100),
            1);

    private static RealtimeTranslationStreamEvent Translation(string text, string eventId) =>
        new(
            RealtimeTranslationLane.Translation(RealtimeTranslationOutputLanguage.English),
            new RealtimeTranslationServerEvent.OutputTranscriptDelta(text, eventId, 200),
            1);
}

File: RealtimeTranslator/Tests/RealtimeSubtitleAssemblerAudioLossTests.swift
import XCTest
@testable import RealtimeTranslator

final class RealtimeSubtitleAssemblerAudioLossTests: XCTestCase {
    private let origin = Date(timeIntervalSince1970: 1_700_000_000)

    func testTaintedSegmentIsNotFinalizedAfterIdle() {
        // Given: 音声欠落を検知して汚染窓を開始する
        var assembler = RealtimeSubtitleAssembler()
        assembler.beginNewEpoch(1)
        assembler.markAudioLoss(now: origin)
        _ = assembler.ingest(source("こんにちは", "s1"), now: origin.addingTimeInterval(1))
        _ = assembler.ingest(translation("Hello", "t1"), now: origin.addingTimeInterval(1))

        // When: idle finalize を越えて評価する
        let update = assembler.tick(now: origin.addingTimeInterval(10))

        // Then: 確定せず、無効化更新で汚染バッファが破棄される
        XCTAssertEqual(update?.isInvalidation, true)
        XCTAssertEqual(update?.shouldFinalize, false)
        XCTAssertEqual(assembler.currentSourceText, "")
        XCTAssertFalse(assembler.isCurrentSegmentTainted)
    }

    func testLossOlderThanWindowAllowsNormalFinalize() {
        // Given: 欠落から8秒を超えて新しいsegmentを開始する
        var assembler = RealtimeSubtitleAssembler()
        assembler.beginNewEpoch(1)
        assembler.markAudioLoss(now: origin)
        _ = assembler.ingest(source("こんにちは", "s1"), now: origin.addingTimeInterval(9))
        _ = assembler.ingest(translation("Hello", "t1"), now: origin.addingTimeInterval(9))

        // When: idle finalize を評価する
        let update = assembler.tick(now: origin.addingTimeInterval(17))

        // Then: 通常どおり確定する
        XCTAssertEqual(update?.shouldFinalize, true)
        XCTAssertEqual(update?.sourceText, "こんにちは")
    }

    func testFinalizedSegmentBeforeLossIsPreserved() {
        // Given: 欠落前にsegmentを確定し、その後に欠落を通知する
        var assembler = RealtimeSubtitleAssembler()
        assembler.beginNewEpoch(1)
        _ = assembler.ingest(source("こんにちは", "s1"), now: origin)
        _ = assembler.ingest(translation("Hello", "t1"), now: origin)
        XCTAssertEqual(assembler.tick(now: origin.addingTimeInterval(8))?.shouldFinalize, true)

        // When: 欠落を通知する
        assembler.markAudioLoss(now: origin.addingTimeInterval(9))

        // Then: 既に確定したsegmentの状態を変更しない
        XCTAssertFalse(assembler.isCurrentSegmentTainted)
        XCTAssertEqual(assembler.currentSourceText, "")
    }

    func testLanguageSwitchDoesNotFinalizeTaintedSegment() {
        // Given: 汚染されたsegmentに原文と訳文を取り込む
        var assembler = RealtimeSubtitleAssembler()
        assembler.beginNewEpoch(1)
        assembler.markAudioLoss(now: origin)
        _ = assembler.ingest(source("こんにちは。", "s1"), now: origin.addingTimeInterval(1))
        _ = assembler.ingest(translation("Hello.", "t1"), now: origin.addingTimeInterval(1))

        // When: 言語切替で分割する
        let split = assembler.splitForLanguageSwitch(at: 6, now: origin.addingTimeInterval(2))

        // Then: 汚染されたprefixは確定しない
        XCTAssertNil(split.finalized)
    }

    func testLanguageSwitchKeepsTaintedSuffixFromFinalizing() {
        // Given: 汚染された原文の途中に言語切替境界がある
        var assembler = RealtimeSubtitleAssembler()
        assembler.beginNewEpoch(1)
        assembler.expectLane(.english)
        assembler.markAudioLoss(now: origin)
        _ = assembler.ingest(
            source("こんにちはHello", "s1"),
            now: origin.addingTimeInterval(1)
        )
        _ = assembler.ingest(
            translation("Hello", "t1"),
            now: origin.addingTimeInterval(1)
        )

        // When: 切替で suffix を残し、新しい訳文が届いて idle する
        let split = assembler.splitForLanguageSwitch(at: 5, now: origin.addingTimeInterval(2))
        XCTAssertNil(split.finalized)
        XCTAssertEqual(assembler.currentSourceText, "Hello")
        XCTAssertTrue(assembler.isCurrentSegmentTainted)

        assembler.expectLane(.japanese)
        _ = assembler.ingest(
            RealtimeTranslationStreamEvent(
                lane: .translation(.japanese),
                event: .outputTranscriptDelta(delta: "こんにちは", eventID: "t2", elapsedMs: nil),
                epoch: 1
            ),
            now: origin.addingTimeInterval(3)
        )
        let update = assembler.tick(now: origin.addingTimeInterval(12))

        // Then: 汚染 suffix は確定せず無効化される
        XCTAssertEqual(update?.isInvalidation, true)
        XCTAssertEqual(update?.shouldFinalize, false)
        XCTAssertEqual(assembler.currentSourceText, "")
        XCTAssertFalse(assembler.isCurrentSegmentTainted)
    }

    private func source(_ value: String, _ id: String) -> RealtimeTranslationStreamEvent {
        RealtimeTranslationStreamEvent(
            lane: .source,
            event: .inputTranscriptDelta(delta: value, eventID: id, elapsedMs: nil),
            epoch: 1
        )
    }

    private func translation(_ value: String, _ id: String) -> RealtimeTranslationStreamEvent {
        RealtimeTranslationStreamEvent(
            lane: .translation(.english),
            event: .outputTranscriptDelta(delta: value, eventID: id, elapsedMs: nil),
            epoch: 1
        )
    }
}
